build/utf8/valid_asm.go

```diff
@@ -0,0 +1,547 @@
+	SUBQ(Imm(32), d)
+	ADDQ(bitset, d)
+	ADDQ(Imm(32), n)
+	SUBQ(bitset, n)
+	JMP(LabelRef("stdlib"))
+
+	Comment("Process one 32B block of data")
+	Label("process")
+	currentBlockY := YMM()
+
+	Comment("Load the next block of bytes")
+	VMOVDQU(Mem{Base: d}, currentBlockY)
+	SUBQ(U8(32), n)
+	ADDQ(U8(32), d)
+
+	Comment("Fast check to see if ASCII")
+	tmp := GP32()
+	VPMOVMSKB(currentBlockY, tmp)
+	CMPL(tmp, Imm(0))
+	JNZ(LabelRef("non_ascii"))
+
+	Comment("If this whole block is ASCII, there is nothing to do, and it is an error if any of the previous code point was incomplete.")
+	VPOR(errorY, incompletePreviousBlockY, errorY)
+	JMP(LabelRef("check_input"))
+
+	Label("non_ascii")
+	Comment("Check errors on the high nibble of the previous byte")
+	previousY := pushLastByteFromAToFrontOfB(previousBlockY, currentBlockY)
+
+	highPrev := highNibbles(previousY, nibbleMaskY)
+	VPSHUFB(highPrev, nibble1Y, highPrev)
+
+	Comment("Check errors on the low nibble of the previous byte")
+	lowPrev := lowNibbles(previousY, nibbleMaskY)
+	VPSHUFB(lowPrev, nibble2Y, lowPrev)
+	VPAND(lowPrev, highPrev, highPrev)
+
+	Comment("Check errors on the high nibble on the current byte")
+	highCurr := highNibbles(currentBlockY, nibbleMaskY)
+	VPSHUFB(highCurr, nibble3Y, highCurr)
+	VPAND(highCurr, highPrev, highPrev)
+
+	Comment("Find 3 bytes continuations")
+	off2 := pushLast2BytesFromAToFrontOfB(previousBlockY, currentBlockY)
+	VPSUBUSB(continuation3BytesY, off2, off2)
+
+	Comment("Find 4 bytes continuations")
+	off3 := pushLast3BytesFromAToFrontOfB(previousBlockY, currentBlockY)
+	VPSUBUSB(continuation4BytesY, off3, off3)
+
+	Comment("Combine them to have all continuations")
+	continuationBitsY := YMM()
+	VPOR(off2, off3, continuationBitsY)
+
+	Comment("Perform a byte-sized signed comparison with zero to turn any non-zero bytes into 0xFF.")
+	tmpY := zeroOutVector(YMM())
+	VPCMPGTB(tmpY, continuationBitsY, continuationBitsY)
+
+	Comment("Find bytes that are continuations by looking at their most significant bit.")
+	msbMask := ConstArray64("msb_mask",
+		0x8080808080808080,
+		0x8080808080808080,
+		0x8080808080808080,
+		0x8080808080808080,
+	)
+
+	msbMaskY := YMM()
+	VMOVDQU(msbMask, msbMaskY)
+	VPAND(msbMaskY, continuationBitsY, continuationBitsY)
+
+	Comment("Find mismatches between expected and actual continuation bytes")
+	VPXOR(continuationBitsY, highPrev, continuationBitsY)
+
+	Comment("Store result in sticky error")
+	VPOR(errorY, continuationBitsY, errorY)
+
+	Comment("Prepare for next iteration")
+	VPSUBUSB(incompleteMaskY, currentBlockY, incompletePreviousBlockY)
+	VMOVDQU(currentBlockY, previousBlockY)
+	MOVB(Imm(1), hasPreviousBlock)
+
+	Comment("End of loop")
+	JMP(LabelRef("check_input"))
+
+	Label("end")
+
+	Comment("If the previous block was incomplete, this is an error.")
+	VPOR(incompletePreviousBlockY, errorY, errorY)
+
+	Comment("Return whether any error bit was set")
+	VPTEST(errorY, errorY)
+	SETEQ(ret.Addr)
+	RET()
```

Comment:
Let's call `VZEROUPPER` before returning since we used YMM registers, otherwise instructions that switch back to using XMM registers take a perf hit.
```suggestion
	VZEROUPPER()
	RET()
```
More details here https://stackoverflow.com/questions/41303780/why-is-this-sse-code-6-times-slower-without-vzeroupper-on-skylake/41349852#41349852

Reply:
I had no idea, thank you!